src/cargo/core/compiler/job_queue.rs

```diff
@@ -617,6 +618,10 @@ impl<'a, 'cfg> JobQueue<'a, 'cfg> {
             Artifact::All => self.timings.unit_finished(id, unlocked),
             Artifact::Metadata => self.timings.unit_rmeta_finished(id, unlocked),
         }
+        if unit.is_std && unit.kind == super::Kind::Target && !cx.bcx.build_config.build_plan {
+            let rmeta = artifact == Artifact::Metadata;
+            standard_lib::add_sysroot_artifact(cx, unit, rmeta)?;
+        }
```

Comment:
Ah I see what you mean. Thinking more on this though, I was actually assuming that we'd just change `--out-dir` for `is_std` crates. It looks like otherwise they're gonna show up both in the libdir as well as the main `deps` folder, right? If we switched `--out-dir` I think it'd solve this issue naturally?

Reply:
Er wait no we already talked about it and ruled it out because we want to clean out the sysroot on each build.

In any case this poses an interesting scenario where the dependencies of `std` can be loaded from either the sysroot or `-Ldependency=.../deps`, they're actually all the same. I'm asking this from the point of view of concurrency which I think this issue is trying to solve (making sure we do this before the next compilation starts), but maybe that's not actually necessary? 

Ok I want to try to work this through here in a coment...

* So for std->std dependencies, as well as non-std->non-std dependencies, those all use `--extern` I believe. If that's the case then I think that nothing is necessary here in terms of sequencing things right because everything is loaded from `deps`
* The only case I think this cares about is non-std -> std dependencies, since they're loaded through the sysroot. Specifically you're worried here that we will kick off compilations that depend on `core`, but nothing may actually be in the sysroot yet, so the compiler won't be able to find anything.

Does that sound right? If so...

So presumably this would otherwise go [somewhere around here](https://github.com/rust-lang/cargo/blob/aa6b7e01abce30091cc594cb23a15c46cead6e24/src/cargo/core/compiler/mod.rs#L1208-L1229) (transitively), and you're right that's a "pretty generic path" which isn't a great spot to put it.

Ok if you agree with what I've said above, let's leave this here, but add a comment perhaps saying that we may want to figure out how to offload it to worker threads eventually?

Reply:
That would be cleaner.  For some reason I was thinking that was going to be harder (since the output directory is somewhat global), but it probably won't be too hard. I'll give it a shot.

Reply:
> we want to clean out the sysroot on each build

Oh, yea, that's the actual reason.  Sorry, I'm mis-remembering things.

> Ok if you agree with what I've said above, let's leave this here, but add a comment perhaps saying that we may want to figure out how to offload it to worker threads eventually?

That makes sense.  I'll add a comment.

I was previously trying to think of a way that `OutputFile::hard_link` (or something like it) could be used, but it still runs into the problem with rmeta files needing to be available, and linking isn't done until after compilation finishes.